fl4health/losses/deep_mmd_loss.py

```diff
@@ -0,0 +1,224 @@
+
+        # Initialize the model
+        self.featurizer = ModelLatentF(input_size, hidden_size, output_size).to(self.device)
+
+        # Initialize parameters
+        self.epsilonOPT: torch.Tensor = torch.log(torch.from_numpy(np.random.rand(1) * 10 ** (-10)).to(self.device))
+        self.epsilonOPT.requires_grad = self.training
+        self.sigmaOPT: torch.Tensor = torch.sqrt(torch.tensor(2 * 32 * 32, dtype=torch.float).to(self.device))
+        self.sigmaOPT.requires_grad = self.training
+        self.sigma0OPT: torch.Tensor = torch.sqrt(torch.tensor(0.005, dtype=torch.float).to(self.device))
+        self.sigma0OPT.requires_grad = self.training
+
+        # Initialize optimizers
+        self.optimizer_F = torch.optim.Adam(
+            list(self.featurizer.parameters()) + [self.epsilonOPT] + [self.sigmaOPT] + [self.sigma0OPT], lr=self.lr
+        )
+
+    def Pdist2(self, x: torch.Tensor, y: Optional[torch.Tensor]) -> torch.Tensor:
+        """compute the paired distance between x and y."""
+        x_norm = (x**2).sum(1).view(-1, 1)
+        if y is not None:
+            y_norm = (y**2).sum(1).view(1, -1)
+        else:
+            y = x
+            y_norm = x_norm.view(1, -1)
+        Pdist = x_norm + y_norm - 2.0 * torch.mm(x, torch.transpose(y, 0, 1))
+        Pdist[Pdist < 0] = 0
+        return Pdist
+
+    def h1_mean_var_gram(
+        self, Kx: torch.Tensor, Ky: torch.Tensor, Kxy: torch.Tensor, is_var_computed: bool, use_1sample_U: bool = True
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], torch.Tensor]:
+        """compute value of MMD and std of MMD using kernel matrix."""
+        Kxxy = torch.cat((Kx, Kxy), 1)
+        Kyxy = torch.cat((Kxy.transpose(0, 1), Ky), 1)
+        Kxyxy = torch.cat((Kxxy, Kyxy), 0)
+        nx = Kx.shape[0]
+        ny = Ky.shape[0]
+        is_unbiased = True
+        if is_unbiased:
+            xx = torch.div((torch.sum(Kx) - torch.sum(torch.diag(Kx))), (nx * (nx - 1)))
+            yy = torch.div((torch.sum(Ky) - torch.sum(torch.diag(Ky))), (ny * (ny - 1)))
+            # one-sample U-statistic.
+            if use_1sample_U:
+                xy = torch.div((torch.sum(Kxy) - torch.sum(torch.diag(Kxy))), (nx * (ny - 1)))
+            else:
+                xy = torch.div(torch.sum(Kxy), (nx * ny))
+            mmd2 = xx - 2 * xy + yy
+        else:
+            xx = torch.div((torch.sum(Kx)), (nx * nx))
+            yy = torch.div((torch.sum(Ky)), (ny * ny))
+            # one-sample U-statistic.
+            if use_1sample_U:
+                xy = torch.div((torch.sum(Kxy)), (nx * ny))
+            else:
+                xy = torch.div(torch.sum(Kxy), (nx * ny))
+            mmd2 = xx - 2 * xy + yy
+        if not is_var_computed:
+            return mmd2, None, Kxyxy
+        hh = Kx + Ky - Kxy - Kxy.transpose(0, 1)
+        V1 = torch.dot(hh.sum(1) / ny, hh.sum(1) / ny) / ny
+        V2 = (hh).sum() / (nx) / nx
+        varEst = 4 * (V1 - V2**2)
+        return mmd2, varEst, Kxyxy
```

Comment:
Correct me if I'm wrong, but in all instances where this function is called, we don't make use of `Kxyxy`. Can we just drop it from the return values?